# GTEx_microbiome_analysis

This is the repository with all the code used in the manuscript 

Machine learning models reveal microbial signatures in healthy human tissues, challenging the sterility of human organs
  Anargyros Skoulakis1,2, Giorgos Skoufos1,2, Armen Ovsepian1,2, Artemis G. Hatzigeorgiou1,2

It contains the code for recreating the results (to do that you should have access to the GTEx data as accession of BAM files of RNA-seq data of samples are restricted). Also, it contains the code for redoing all the analyses presented in the paper, using the final results and metadata, saved in Zenodo (link https://zenodo.org/uploads/10980664). Using the final results in Zenodo and the publicly available metadata of samples, one can replicate the results. Only the analysis for discriminating the phenotypic traits and medical history of samples demands the access to the full metadata file that need access to GTEx data. Lastly, as the publicly available metadata have different headers of the full metadata, the scripts should be modified accordingly.

In the folder code there is the scripts for creating all the necessary files for the analysis!
